### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] False positive in 8.59.0 from discriminated union handling

### ISSUE THREAD

**Issue by joshkel** [typescript-eslint/typescript-eslint#12277]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&fileType=.tsx&code=JYOwLgpgTgZghgYwgAgPIAczAPYgM4CCyA3sjNtgFzJ5hSgDmA3MgL4BQoksiKGWuPACESyAEZwo1EAFcAtmOgsO7MAE90KAiDWoxAKwgIwyALztkyAD6j1m6gHI4Dltkw58Afmr8PhNhbWthoQjmIuyG4CXj7ugiKsTOzsACZGADaSKAiCJtgGAIzU2roGRmBJaQiZUNm5kQYATMU6eobGlRlZZDIgxh7IMngQABT5%2Bi2l7WAAlNQAbtjAKUnsAPRryAAqIQDKCPSYyAhwICDYJiCSUNgA7sgpwHgHwHKgcJApgyB%2ByLcAFhAQMdsHJ0JJGMgwLdsN8BvM4OkZBA8AA6TgwZBjQqouwoUwEhr6Rq4kIzEiBDbIACiuwAMlxkKAclBasZ0mpkLV0NgoGA8NQAERbf5PZBwPDDPkDMW9EBGFF4SRqQWUzZCGQmWpybDzSGCiXinSC45wIYo7a7RoAZkajUC5pGpFRLvGBQANJE4vhqOMSVFfobTpzWDMkhwgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```TypeScript
interface OptionsA { foo: string; }
interface OptionsB { bar: number; }

type AnyObject =
  | { type: 'a'; options?: OptionsA }
  | { type: 'b'; options?: OptionsB };

declare const obj1: AnyObject;
declare const obj2: AnyObject;
declare function use(obj: AnyObject): void;

// TypeScript cannot narrow discriminated unions when comparing two union values.
if (obj1.type === obj2.type) {
  // ESLint incorrectly reports: "This assertion is unnecessary"
  // But removing "as any" causes TS2322
  use({ ...obj1, options: obj2.options as any });
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

No typescript-eslint errors, since the assertion is necessary for TypeScript to compile.

### Actual Result

"This assertion is unnecessary since the receiver accepts the original type of the expression."

**Comment by bradzacher** [comment]
This is a funny case. First of all the code is impossible to write without either duplication or assertions. For example even the code `{...obj1, options: obj1.options}` does not work even though logically it works.

So there's some complexity here already with the types...

Now the rule isn't technically wrong here. The type of the argument's `options` property is in fact `OptionsA | OptionsB` - which matches the exact type of `obj1.options`.

The thing the rule misses is the fact that the final result of the object expression is `{ options: OptionsA | OptionsB | undefined; type: "a"; } | { options: OptionsA | OptionsB | undefined; type: "b"; }`, which is subtly different to `OptionsA | OptionsB`, even if `typeof foo.options` is the same in both cases.

I don't know if this is truly solvable, TBH.

**Comment by StyleShit** [comment]
I just encountered a very similar bug with storybook stories ([repro](https://typescript-eslint.io/play/#ts=5.9.2&showAST=es&fileType=.tsx&code=C4TwDgpgBACgTgezAZygXgFBSgHygby2ygFsBXAG2AEswKIB%2BALigDMBDC5CAbiOwBunMhBbJgcagDsA5n2wBfInkLFSlGnVFQJI%2BcSEURYidJkBtALr6FfDABMIAYwrs40VmSlOaCKVABlCHofAAowRBQWeCRkAEo7T29ff3cpRzhQtxlkaMj4giJ3YDI4fwAeIJDgdSpaejR8bOQAOnI6rSh2VCkIAQg4BShDEUbmlpHobqhe-sGoAHoAPj4FIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false))

> I don't know if this is truly solvable, TBH.

I agree

Not necessarily the best idea, but maybe we could exclude `any` / `never` from this (unless the resolved type is actually `any` / `never`), because they're usually used in such cases?

I know this is a very wide assumption here (that might cause false negatives), but... 🤷
